pkg/proxy/iptables/proxier.go

```diff
@@ -572,45 +590,56 @@ func (proxier *Proxier) OnEndpointsSynced() {
 	proxier.syncProxyRules(syncReasonEndpoints)
 }
 
-// Convert a slice of api.Endpoints objects into a map of service-port -> endpoints.
-func buildNewEndpointsMap(allEndpoints endpointsMap, curMap proxyEndpointMap, hostname string) (newMap proxyEndpointMap, hcEndpoints map[types.NamespacedName]int, staleSet map[endpointServicePair]bool) {
+func (proxier *Proxier) mergeEndpointsMap(endpointsMap proxyEndpointsMap) {
+	for svcPort := range endpointsMap {
+		proxier.endpointsMap[svcPort] = endpointsMap[svcPort]
+	}
 
-	// return values
-	newMap = make(proxyEndpointMap)
-	hcEndpoints = make(map[types.NamespacedName]int)
-	staleSet = make(map[endpointServicePair]bool)
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ExternalTrafficLocalOnly) {
+		return
+	}
+	localIPs := getLocalIPs(endpointsMap)
+	for nsn, ips := range localIPs {
+		proxier.hcEndpoints[nsn] = len(ips)
+	}
+}
 
-	// Update endpoints for services.
-	for _, endpoints := range allEndpoints {
-		accumulateEndpointsMap(endpoints, hostname, &newMap)
+func (proxier *Proxier) unmergeEndpointsMap(endpointsMap proxyEndpointsMap) {
+	for svcPort := range endpointsMap {
+		delete(proxier.endpointsMap, svcPort)
 	}
-	// Check stale connections against endpoints missing from the update.
-	// TODO: we should really only mark a connection stale if the proto was UDP
-	// and the (ip, port, proto) was removed from the endpoints.
-	for svcPort, epList := range curMap {
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ExternalTrafficLocalOnly) {
+		return
+	}
+	localIPs := getLocalIPs(endpointsMap)
+	for nsn := range localIPs {
+		delete(proxier.hcEndpoints, nsn)
+	}
+}
+
+func (proxier *Proxier) detectStaleConnections(oldEndpointsMap, newEndpointsMap proxyEndpointsMap) {
```

Comment:
I agree with maing this a function.

I just want to clarify with your second suggestion. I think there are two possibilities here:
1. make "newEndpoints" a "full" copy (but that's bad from performance reasons to always make a copy of it)
2. make "newEndpoitns" only accumulate changes (i.e. only storing desired state).

The (2) should work (the only tricky part is how to store "deleted" state, but I guess we can just store nil in a map and having a nil for a given key is different than having nothing for a key, so that should work.
With that we will be applying those changes similarly to what I'm doing now.

If that was your suggestion, then I agree it makes perfect sense, and I'm happy to change it. (let me know if I misunderstood something).